src/vos/vos_io.c

```diff
@@ -1868,16 +1896,24 @@ dkey_update(struct vos_io_context *ioc, uint32_t pm_ver, daos_key_t *dkey,
 	daos_handle_t		 ak_toh;
 	struct vos_krec_df	*krec;
 	uint32_t		 update_cond = 0;
+	uint32_t                 flags         = SUBTR_CREATE;
 	bool			 subtr_created = false;
 	int			 i, rc;
 
 	rc = obj_tree_init(obj);
 	if (rc != 0)
 		return rc;
 
-	rc = key_tree_prepare(obj, obj->obj_toh, VOS_BTR_DKEY, dkey,
-			      SUBTR_CREATE, DAOS_INTENT_UPDATE, &krec, &ak_toh,
-			      ioc->ic_ts_set);
+	if (vos_obj_flat_kv_supported(ioc->ic_cont, obj->obj_id)) {
+		if (ioc->ic_iod_nr == 1) {
+			flags |= SUBTR_FLAT;
+			if (ioc->ic_iods[0].iod_type == DAOS_IOD_ARRAY)
+				flags |= SUBTR_EVT;
+		}
+	}
+
+	rc = key_tree_prepare(obj, obj->obj_toh, VOS_BTR_DKEY, dkey, flags, DAOS_INTENT_UPDATE,
+			      &krec, &ak_toh, ioc->ic_ts_set);
```

Comment:
I wonder should we skip dkey as well? i.e. attaching the ev tree to object initially, and only add dkey tree only if there are multiple dkeys being insert.

Reply:
That's tricky because we need to store some dkey specific information such as the incarnation log and the actual key.  The most natural place for that is in the actual dkey.

I think DAOS-13560 gets us most of the way to what your are saying in that it would avoid creating all of the tree nodes in such a case.